src/gui/src/clockWidget.cpp

```diff
@@ -0,0 +1,1202 @@
+  }
+
+  for (auto* driver : drivers) {
+    driver->setupToolTip();
+  }
+  for (auto* fan : fanout) {
+    fan->setupToolTip();
+  }
+
+  return drivers;
+}
+
+ClockNodeGraphicsViewItem* ClockTreeView::addRootToScene(
+    qreal x,
+    const PinArrival& output_pin,
+    sta::dbNetwork* network)
+{
+  odb::dbITerm* iterm;
+  odb::dbBTerm* bterm;
+  network->staToDb(output_pin.pin, iterm, bterm);
+
+  ClockNodeGraphicsViewItem* node = nullptr;
+  if (iterm != nullptr) {
+    node = new ClockRootNodeGraphicsViewItem(iterm);
+  } else {
+    node = new ClockRootNodeGraphicsViewItem(bterm);
+  }
+
+  node->setPos(x, convertDelayToY(output_pin.delay));
+  scene_->addItem(node);
+
+  node->setExtraToolTip("Launch: " + convertDelayToString(output_pin.delay));
+
+  return node;
+}
+
+ClockNodeGraphicsViewItem* ClockTreeView::addLeafToScene(
+    qreal x,
+    const PinArrival& input_pin,
+    sta::dbNetwork* network)
+{
+  odb::dbITerm* iterm;
+  odb::dbBTerm* bterm;
+  network->staToDb(input_pin.pin, iterm, bterm);
+
+  ClockRegisterNodeGraphicsViewItem* node
+      = new ClockRegisterNodeGraphicsViewItem(iterm);
+  node->scaleSize(leaf_scale_);
+
+  node->setPos({x, convertDelayToY(input_pin.delay)});
+  node->setExtraToolTip("Arrival: " + convertDelayToString(input_pin.delay));
+  scene_->addItem(node);
+
+  connect(node->getHighlightAction(), &QAction::triggered, [this, iterm]() {
+    emit highlightTo(iterm);
+  });
+
+  return node;
+}
+
+ClockNodeGraphicsViewItem* ClockTreeView::addBufferToScene(
+    qreal x,
+    const PinArrival& input_pin,
+    const PinArrival& output_pin,
+    sta::dbNetwork* network)
+{
+  auto convert_pin = [&network](sta::Pin* pin) -> odb::dbITerm* {
+    odb::dbITerm* iterm;
+    odb::dbBTerm* bterm;
+    network->staToDb(pin, iterm, bterm);
+    return iterm;
+  };
+
+  odb::dbITerm* input_term = convert_pin(input_pin.pin);
+  odb::dbITerm* output_term = convert_pin(output_pin.pin);
+
+  const qreal delay_y
+      = convertDelayToY(output_pin.delay) - convertDelayToY(input_pin.delay);
+
+  ClockNodeGraphicsViewItem* node
+      = new ClockBufferNodeGraphicsViewItem(input_term, output_term, delay_y);
+  node->setPos({x, convertDelayToY(input_pin.delay)});
+  scene_->addItem(node);
+
+  QString tooltip;
+  tooltip += "Input: " + ClockNodeGraphicsViewItem::getITermName(input_term);
+  tooltip += "\n";
+  tooltip += "Input arrival: " + convertDelayToString(input_pin.delay);
+  tooltip += "\n";
+  tooltip += "Output: " + ClockNodeGraphicsViewItem::getITermName(output_term);
+  tooltip += "\n";
+  tooltip += "Output launch: " + convertDelayToString(output_pin.delay);
+  node->setExtraToolTip(tooltip);
+
+  return node;
+}
+
+void ClockTreeView::highlightTo(odb::dbITerm* term)
+{
+  enableRenderer(true);
+  scene_->setClearPathEnable(true);
+
+  renderer_->setPathTo(term);
+}
+
+void ClockTreeView::clearHighlightTo()
+{
+  renderer_->clearPathTo();
+}
+
+void ClockTreeView::save(const QString& path)
```

Comment:
Is this exposed in tcl?

Reply:
not yet, I wanted to get the rest of things squared away before adding a tcl interface for it.

Reply:
added save_clocktree_image to tcl to be able to save the images